crates/bevy_render/src/batching/gpu_preprocessing.rs

```diff
@@ -76,10 +129,114 @@ pub struct PreprocessWorkItem {
     /// The index of the batch input data in the input buffer that the shader
     /// reads from.
     pub input_index: u32,
-    /// The index of the `MeshUniform` in the output buffer that we write to.
+    /// In direct mode, this is the index of the `MeshUniform` in the output
+    /// buffer that we write to. In indirect mode, this is the index of the
+    /// [`IndirectParameters`].
     pub output_index: u32,
 }
 
+/// The `wgpu` indirect parameters structure.
+///
+/// This is actually a union of the two following structures:
+///
+/// ```
+/// #[repr(C)]
+/// struct ArrayIndirectParameters {
+///     vertex_count: u32,
+///     instance_count: u32,
+///     first_vertex: u32,
+///     first_instance: u32,
+/// }
+///
+/// #[repr(C)]
+/// struct ElementIndirectParameters {
+///     index_count: u32,
+///     instance_count: u32,
+///     first_vertex: u32,
+///     base_vertex: u32,
+///     first_instance: u32,
+/// }
+/// ```
+///
+/// We actually generally treat these two variants identically in code. To do
+/// that, we make the following two observations:
+///
+/// 1. `instance_count` is in the same place in both structures. So we can
+/// access it regardless of the structure we're looking at.
+///
+/// 2. The second structure is one word larger than the first. Thus we need to
+/// pad out the first structure by one word in order to place both structures in
+/// an array. If we pad out `ArrayIndirectParameters` by copying the
+/// `first_instance` field into the padding, then the resulting union structure
+/// will always have a read-only copy of `first_instance` in the final word. We
+/// take advantage of this in the shader to eliminate branching.
+#[derive(Clone, Copy, Pod, Zeroable, ShaderType)]
+#[repr(C)]
+pub struct IndirectParameters {
+    /// For `ArrayIndirectParameters`, `vertex_count`; for
+    /// `ElementIndirectParameters`, `index_count`.
+    pub data0: u32,
```

Comment:
Maybe we could call this:
```suggestion
    pub vertex_or_index_count: u32,
```

Reply:
Done.